### PR TITLE
chore: Switch discovery document updates from app engine cron to github action

### PR DESCRIPTION
**Pull request by dazuma**
Replaces #5305.

We're trying to turn down this repo's production services and delete as much of it as possible. Currently, the live parts of this repo are:

1. A cache for discovery documents (in the `/discoveries` directory) which is used by apiary generators for some languages
2. An App Engine cron and a GitHub Action auto-approver script that updates the discovery document cache hourly
3. Some code (in the `/src` and `/toolkit` directories) that is referenced from a google3 script that ACTools occasionally runs to manually update apiary samples

This PR is one step toward turning stuff down. It reimplements part (2) above in a bit of Python code that runs in an hourly scheduled GitHub Action. There are two python scripts: one that downloads the current discovery documents and updates the `/discoveries` directory, and another that opens a pull request, auto-approves it, and sets automerge.

This replaces the last live function of the App Engine app, which we can then turn down along with removing its attendant autoapprover script. (This PR just disables the associated App Engine crons; we will delete the entire `/server` directory in a subsequent PR. Also note that we are reimplementing the auto-approve functionality inside the python script rather than keeping it in a separate GitHub Action; this should make the system a bit more robust.)

### CODE REVIEW

**Comment by dazuma** [comment]
@SurferJeffAtGoogle PTAL

**Comment by dazuma** [comment]
Note: please review but do not submit. I need to coordinate submission with disabling the app engine app.

File: scripts/open_pr.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    if os.getenv(MAIN_TOKEN_ENV) == None:
+        sys.exit("GITHUB_TOKEN environment variable must be set")
+    ensure_git_identity()
+    ensure_git_fork()
+
+
+def ensure_git_identity():
+    """Ensure that the git identity (name and email) is set"""
+    if subprocess.run(["git", "config", "--get", "user.name"], capture_output=True).returncode != 0:
+        logging.info("Need to set git user.name.")
```
Comment (SurferJeffAtGoogle) [inline scripts/open_pr.py:81]: I first interpreted this log message to mean that I forget to set the user.name, and I need to set it.  Would you mind changing it to `f"Setting git user.name to {GIT_USER_NAME}."`

File: scripts/open_pr.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        subprocess.run(["git", "config", "--local", "user.name", GIT_USER_NAME])
+    if subprocess.run(["git", "config", "--get", "user.email"], capture_output=True).returncode != 0:
+        logging.info("Need to set git user.email.")
```
Comment (SurferJeffAtGoogle) [inline scripts/open_pr.py:84]: ditto

File: scripts/open_pr.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    subprocess.run(["gh", "repo", "sync", fork_repo_name], check=True)
+    if subprocess.run(["git", "remote", "get-url", REMOTE_NAME], capture_output=True).returncode != 0:
+        logging.info(f"Need to create the remote {REMOTE_NAME}.")
+        token = os.getenv(MAIN_TOKEN_ENV)
+        subprocess.run(["git", "remote", "add", REMOTE_NAME, f"https://{username}:{token}@github.com/{fork_repo_name}.git"], check=True)
+
+
+def open_pr():
+    """Actually open the pull request"""
+    branch = commit_changes()
```
Comment (SurferJeffAtGoogle) [inline scripts/open_pr.py:102]: `commit_changes()` takes one argument.  I expect this line would raise an exception.

File: scripts/update_disco.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            f.seek(0)
+            f.truncate()
+            f.write(document.content)
+
+
+def update_index(index_document):
+    """Updates the index file in the discoveries dir
+
+    Arguments:
+        index_document {DocumentInfo} -- The index
```
Comment (SurferJeffAtGoogle) [inline scripts/update_disco.py:181]: With python3, we can use type annotations in the code.  Then, [mypy](https://github.com/python/mypy) can find lots of bugs before the code runs.

For Github automation tools, we run mypy and black on all our Python code.

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: scripts/open_pr.py (hunk 1 above)
Comment (dazuma) [reply scripts/open_pr.py]: Done

File: scripts/open_pr.py (hunk 2 above)
Comment (dazuma) [reply scripts/open_pr.py]: Done

File: scripts/open_pr.py (hunk 3 above)
Comment (dazuma) [reply scripts/open_pr.py]: Thanks, I failed to retest after doing a last refactor. Fixed.

File: scripts/update_disco.py (hunk 4 above)
Comment (dazuma) [reply scripts/update_disco.py:210]: I went and added type annotations, ran both mypy and black, and retested end-to-end. Also added a CI workflow that runs mypy and black --check on these scripts.

File: scripts/update_disco.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        if existing_content == index_document.content:
+            logging.info(f"Index file {filename} is unchanged")
+        else:
+            print(f"UPDATING index file {filename}")
+            f.seek(0)
+            f.truncate()
+            f.write(index_document.content)
+
+
+main()
```
Comment (SurferJeffAtGoogle) [inline scripts/update_disco.py:224]: Add an `if __name__ == '__main__'`: here.

Then, write tests that cover all the code in `update_disco.py` that doesn't make network calls.  That looks like all the code except the `main()` and `load_index()` functions.

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: scripts/update_disco.py (hunk 5 above)
Comment (dazuma) [reply scripts/update_disco.py]: Done

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Pull request merged**